### PR TITLE
kubeadm: Add --cert-dir, --cert-altnames instead of --api-external-dns-names

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

 - For the beta kubeadm init UX, we need this change
 - Also adds the `kubeadm phase certs selfsign` command that makes the phase invokable independently

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

This PR depends on https://github.com/kubernetes/kubernetes/pull/41897

**Release note**:

```release-note
```
@dmmcquay @pires @jbeda @errordeveloper @mikedanese @deads2k @liggitt

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 1]
```diff
@@ -40,11 +40,12 @@ type MasterConfiguration struct {
 	KubernetesVersion string
 	CloudProvider     string
 	AuthorizationMode string
+	CertAltNames      []string
```
Comment (jbeda) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: We should start documenting these in this file like other APIs.

We can also explicitly list here which phases this input is used for.  May be easier to grok vs. looking at the doc.go files in each package.

CertAltNames: "A list of extra names to include in the Subject Alternate Name section of the API server certificate.  Use this if you will be accessing the server using some DNS name that isn't the host name of the node."

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 2]
```diff
@@ -40,11 +40,12 @@ type MasterConfiguration struct {
...
+	CertificatesDir   string
```
Comment (jbeda) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: "The directory that will hold all of the generated certificates and private keys."

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 1 above)
Comment (jbeda) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: x509 certificate Subject Alternate Name extensions are actually typed.  You can include DNS names, URIs and IP addresses.  I see below that you auto-detect if it is a DNS name or IP address. 

We probably should reflect that typing explicitly.  Either with separate lists for IPs and DNS or by having some sort of prefix here: `dns:foo.com`, `ip:x.x.x.x`.  I'd vote for separate parameters.  `CertExtraDNS` and `CertExtraIP`.

We should then document that these are added to the "Subject Alternate Name" section of the certificate.

But -- we are down to the wire so I won't insist.  But if we want to do more than DNS or IP addresses we'll have to add more parameters and be more explicit.

File: cmd/kubeadm/app/phases/certs/doc.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ package certs
 	INPUTS:
 		From MasterConfiguration
 			.API.AdvertiseAddresses is needed for knowing which IPs the certs should be signed for
-			.API.ExternalDNSNames is needed for knowing which DNS names the certs should be signed for
+			AltNames is needed for knowing which DNS names the certs should be signed for
```
Comment (jbeda) [inline cmd/kubeadm/app/phases/certs/doc.go]: Update this to match -- CertAltNames.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		Short: "Generate the CA, APIServer signing/client cert, the ServiceAccount public/private keys and a CA and client cert for the front proxy",
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunSelfSign(config)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	cmd.Flags().StringVar(&config.Networking.DNSDomain, "dns-domain", kubeadmapiext.DefaultServiceDNSDomain, "The DNS Domain for the Kubernetes cluster.")
+	cmd.Flags().StringVar(&config.CertificatesDir, "cert-dir", kubeadmapiext.DefaultCertificatesDir, "The path where to save and store the certificates")
+	cmd.Flags().StringVar(&config.Networking.ServiceSubnet, "service-cidr", kubeadmapiext.DefaultServicesSubnet, "The subnet for the Services in the cluster.")
+	cmd.Flags().StringSliceVar(&config.CertAltNames, "cert-altnames", []string{}, "Optional extra altnames to use for the API Server serving cert. Can be both IP addresses and dns names.")
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/phases/certs.go]: I'd prefer we use the term "Subject Alternative Names" used in description as it is then clear how it exactly relates to the x.509 cert.

**Review by jbeda — APPROVED** [review]
Mostly documentation and nits.  Can be fixed later if need be and I don't want to hold things up.

**Comment by jbeda** [comment]
/lgtm
/approve

As you rebase, you might want to fix some of my comments.

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 1 above)
Comment (pires) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: Agreed.

**Review by pires — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
@jbeda PTAL at this once again. I have some small things to fix here yet and this need some other PR merged so adding do-not-merge for now, but it's ready for review and can probably be lgtm'd again.

File: cmd/kubeadm/app/phases/certs/doc.go (hunk 3 above)
Comment (jbeda) [inline cmd/kubeadm/app/phases/certs/doc.go]: Still not done.

**Review by jbeda — APPROVED** [review]
I still think we should be explicit on IP vs. DNS name for input to alt names but I won't stand on it.  Also need to fix up some of the documentation.

But I'll approve it as we need to get it in.

**Comment by jbeda** [comment]
/lgtm
Need's rebase

**Comment by jbeda** [comment]
@luxas Looks like bazel build failed with relevant error.

**Comment by jbeda** [comment]
@luxas Looks like bazel build failed with relevant error.

**Comment by calebamiles** [comment]
@k8s-bot verify test this
@k8s-bot unit test this

**Comment by calebamiles** [comment]
@ethernetdan, this PR had `lgtm` before code freeze removing `do-not-merge` label. @luxas please rebase. Thanks!

**Comment by luxas** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot gci gke e2e test this

**Comment by luxas** [comment]
This PR is rebased and green now.

Reapplying the lgtm given above

**Comment by luxas** [comment]
@k8s-bot cvm gce e2e test this

**Comment by luxas** [comment]
@k8s-bot gci gke e2e test this

**Pull request merged**